client/app/classView/commentListWidget/commentListWidget.controller.js

```diff
@@ -0,0 +1,148 @@
+        }
+      } else {
+        $scope.comments.push(comment);
+        sortComments();
+      }
+    });
+
+    $scope.comments = [];
+    $scope.lookupTable = {};
+    $scope.$on('broadcastNewRecording', function (event, recordings) {
+      $scope.searchResults = recordings;
+      update();
+    });
+
+    // updates everything
+    function update() {
+      var commentsPromises = $scope.searchResults.map(recording => Comment.query({recording: recording.id}).$promise);
+      $q.all(commentsPromises)
+        .then(recordingComments => {
+          $scope.comments = _.flatten(recordingComments);
+          sortComments();
+          $scope.lookupTable = createLookupTable($scope.comments);
+          changeComment(-1);
+        }
+        );
+    }
+    update();
+
+    // keep track of the currently playing recording
+    $scope.activeRecording = null;
+    $scope.$on('playingRecording', function (event, recording) {
+      $scope.activeRecording = recording.id;
+      updateComment();
+    });
+
+    // keep track of the time into the class
+    $scope.currentTime = new Date(0);
+    $scope.$on('broadcastCurrentTime', function (event, currentTime) {
+      $scope.currentTime = currentTime;
+      updateComment();
+    });
+
+    // checks whether we need to show a new comment
+    $scope.activeCommentIndex = null;
+    function updateComment() {
+      if (!$scope.autoScroll)
+        return;
+
+      // find the best comment to show
+      var bestCommentIndex = null;
+      for (var i = 0; i < $scope.comments.length; i++) {
+        if (new Date($scope.comments[i].time) > new Date($scope.currentTime))
+          break;
+        if ($scope.comments[i].recording == $scope.activeRecording)
+          bestCommentIndex = i;
+      }
+
+      changeComment(bestCommentIndex);
+    }
+
+    // change the comment widget to show the comment at the specified index
+    // pass in -1 to create a new comment
+    function changeComment(index) {
+      if (index == null) // || index == $scope.activeCommentIndex)
+        return;
+
+      $scope.activeCommentIndex = index;
+      var newComment = (index >= 0) ? $scope.comments[index] : null;
+      var newTime = (index >= 0) ? newComment.time : $scope.currentTime;
+      var recording = $scope.activeRecording;
+      if (!recording) {
+        // TODO: better null checks on recording
+        recording = $scope.searchResults[0].id;
+      }
+      $scope.$emit('changeComment', {
+        comment: newComment,
+        time: newTime,
+        recording: recording,
+        lookupTable: $scope.lookupTable
+      });
+    }
+
+    // creates a lookup table from user.id to user.name
+    function createLookupTable(comments) {
+      var result = {};
+      for (var i = 0; i < comments.length; i++) {
+        var poster = comments[i].poster;
+        if (comments[i].poster) {
+          result[poster.id] = poster.firstName + ' ' + poster.lastName;
+        }
+      }
+      return result;
+    }
```

Comment:
Can make this a bit tidier:
return comments.reduce(function (result, comment) {
     if (comment.poster) {
          result[comment.poster.id] = `${poster.firstName} ${poster.lastName}`;
     }
     return result;
}, {});